### PR TITLE
For bergamot-translator into this repostiory as inference-engine

### PR DESCRIPTION
**Pull request by nordzilla**
This PR forks the https://github.com/browsermt/bergamot-translator repository into this repository, self-contained within the `inference-engine` directory. 

It also modifies it in the following ways:
* Remove immediately unnecessary-for-this-project code.
* Refactor and add scripts for build/test tasks.
* Make tasks runnable in docker, listed in `Taskfile.yaml`
* Modify the CODEOWNERS file for this repository.

### CODE REVIEW

**Comment by gregtatum** [comment]
Locally the commits look like this: 

```
c62bea0 2024-09-25 Erik Nordin (HEAD -> fork-bergamot, origin/fork-bergamot) Add review groups to CODEOWNERS
07e3216 2024-09-26 Erik Nordin Move build-wasm script to inference-engine/scripts directory
cf23bf7 2024-09-26 Erik Nordin Add clean script to inference-engine
bb47eed 2024-09-26 Erik Nordin Add unit-tests script to inference-engine
00f4a30 2024-09-26 Erik Nordin Add build-local script to inference-engine
b6906d9 2024-09-26 Erik Nordin Remove unneded doc code
1019fdd 2024-09-26 Erik Nordin Remove unneeded CLI code
27e85d2 2024-09-20 Erik Nordin Remove unneeded Python code
37d0113 2024-09-20 Erik Nordin Remove .circleci and .github files
3da08c9 2024-09-26 Erik Nordin Remove bergamot-translator-tests dependency
cad3963 2024-09-19 Erik Nordin Rename inference-engine/3rd_party/marian-nmt
bbb8442 2024-09-19 Erik Nordin Move inference-engine git submodules to the repository root
285cf6b 2024-09-26 Erik Nordin Add 'inference-engine/' from commit '9271618ebbdc5d21ac4dc4df9e72beb7ce644774'
77479b3 2024-09-24 Greg Tatum (origin/main, origin/HEAD, main) Update the statistics class to use data attributes and use Statistics in merge-mono.py (#853)
```

Then the merge commit looks like:

```
➤ g show 285cf6b
commit 285cf6ba4d1ff96e018944bf5abf348a3945a873
Merge: 77479b3 9271618
Author: Erik Nordin <enordin@mozilla.com>
Date:   Thu Sep 26 12:54:32 2024 -0500

    Add 'inference-engine/' from commit '9271618ebbdc5d21ac4dc4df9e72beb7ce644774'

    git-subtree-dir: inference-engine
    git-subtree-mainline: 77479b3a77703a281bdf8313e168ecf697a376e5
    git-subtree-split: 9271618ebbdc5d21ac4dc4df9e72beb7ce644774
```

File: inference-engine/scripts/build-local.sh
Comment (gregtatum) [inline inference-engine/scripts/build-local.sh:1]: Thought: The general rule of thumb is that if a script starts to get complicated, it should be written in python. These seem to be fairly simple, and are following the practices of the existing repo being ported over, so I think we're still fine to keep these as bash.

**Review by gregtatum — APPROVED** [review]
Looks good to me. I manually scanned through every file, did a clean git clone, and initiated the git submodules. I didn't do a full build using the task commands as I'm currently working on docker stuff, but I figure if it's broken we can follow-up with fixes.

**Comment by gregtatum** [comment]
Let's make sure to get @eu9ene's approval on this before landing.

File: inference-engine/scripts/build-local.sh
Comment (nordzilla) [reply inference-engine/scripts/build-local.sh:1]: I'd be happy to convert them to Python!

It sounds nicer anyway. I kept them as bash to match the pre-existing `build-wasm.sh`.

But after the merge, I wouldn't mind doing a bash-to-Python cleanup. 

I would rather have a Python ecosystem.

File: .github/CODEOWNERS
Comment (eu9ene) [inline .github/CODEOWNERS:24]: Can we call it just "inference" to make it shorter?

File: .gitmodules
Comment (eu9ene) [inline .gitmodules:10]: We already have this module in the root, can we reuse it somehow? Also, it is possible to have all the 3rd party modules in the root?

File: Taskfile.yml
Comment (eu9ene) [inline Taskfile.yml:78]: same here, can make it shorter by replacing inference-engine with inference

File: inference-engine/.clang-format
Comment (eu9ene) [inline inference-engine/.clang-format:1]: All those "." files likely should be in the root of the repo. For example, if we add some more C++ in a separate folder we would reuse those formatting rules.

File: inference-engine/.gitignore
Comment (eu9ene) [inline inference-engine/.gitignore:1]: we already have .gitignore in the root, we can merge this one there

File: inference-engine/LICENSE
Comment (eu9ene) [inline inference-engine/LICENSE:1]: we have a license in the root, should we remove this one?

File: inference-engine/MANIFEST.in
Comment (eu9ene) [inline inference-engine/MANIFEST.in:1]: not sure what this file is for

File: inference-engine/README.md
Comment (eu9ene) [inline inference-engine/README.md:3]: we don't have any circle ci connected here. Let's clean up this readme

File: inference-engine/Doxyfile.in
Comment (eu9ene) [inline inference-engine/Doxyfile.in:1]: We likely won't use doxygen as we have a separate documentation engine based on markdown in /docs (correct me if I'm wrong here). Should we remove this?

File: inference-engine/BERGAMOT_VERSION
Comment (eu9ene) [inline inference-engine/BERGAMOT_VERSION:1]: Do we need this version? Who will maintain it?

File: inference-engine/README.md
Comment (eu9ene) [inline inference-engine/README.md:1]: Having local readme in the directory can be useful but all those building instructions should be moved to /docs/inference.md

File: inference-engine/examples/run-native.sh
Comment (eu9ene) [inline inference-engine/examples/run-native.sh:1]: This is an old example. It should use our models from `firefox-translations-models` They are already in bergamot-translator format. Also, I suggest we move this example to the docs.

File: inference-engine/scripts/build-local.sh
Comment (eu9ene) [reply inference-engine/scripts/build-local.sh:1]: This is a build script and it's pretty much an industry standard to write such scripts in bash. I suggest we leave them alone unless there is some really complex logic and it's getting hard to maintain them.

File: inference-engine/wasm/README.md
Comment (eu9ene) [inline inference-engine/wasm/README.md:1]: another readme that should be moved to docs

**Review by eu9ene — CHANGES_REQUESTED** [review]
There's quite a lot of code. I didn't look at it much but I hope we'll be able to delete everything except the wasm bindings and related classes.

There are quite a lot of small structural suggestions that should be addressed before we can merge to main. Since we decided to share documentation and lining infrastructure we should unify all this right away as it was one of the reasons why we're moving it to this repo.

File: .github/CODEOWNERS
Comment (nordzilla) [reply .github/CODEOWNERS:24]: Sounds good to me.

File: .gitmodules
Comment (nordzilla) [reply .gitmodules:10]: I actually wrote a long comment about this, and I think it got lost in one of my rebases (since these subtree-merge rebases get complicated).

```
# This is the same dependency and repository as `3rd_party/browsermt-marian-dev` below.
#
# When forking `inference-engine` into to this project, I made an earnest attempt to utilize the preexisting
# `3rd_party/browsermt-marian-dev` submodule within `inference-engine`. Unfortunately, I ran into several roadblocks:
#
#   1) I cannot directly add `3rd_party/browsermt-marian-dev` as a cmake subdirectory because cmake is aware that
#      this path is not a subdirectory of the `inference-engine` project root.
#
#   2) Symbolic links do not appear to work for git submodule direcotires the way that they do for regular directories.
#      Even if the symbolic link had linked correctly, it may have still failed due to the considerations of 1).
#
#   3) I tried using cmake to copy the files from `3rd_party/browsermt-marian-dev` into `inference-engine/3rd_party/browsermt-marian-dev`
#      at build time, which would ensure that there is no duplicate reference to the URL in this file, however the upstream dependency itself
#      has hard-coded expectations that the `.git` directory is only one level up, which appears to work correctly for the way git submodules are
#      configured, but does not work if the files are copied over to a regular directory deeper in the repository's directory tree.
#
# It may be possible to remove `3rd_party/browsermt-marian-dev` to instead use `inference-engine/3rd-party/browsermt-marian-dev` everywhere
# within this repository, but I will leave that for a future commit if there is a need to do so.
[submodule "inference-engine/3rd_party/browsermt-marian-dev"]
path = inference-engine/3rd_party/browsermt-marian-dev
url = https://github.com/browsermt/marian-dev
```

I'll be sure to add the comment back in.

File: inference-engine/.clang-format
Comment (nordzilla) [reply inference-engine/.clang-format:1]: Sure. I just kept them here since this was the only directory that had C++ code in it like this, but I'm happy to move these to the root-level directory.

File: inference-engine/.gitignore
Comment (nordzilla) [reply inference-engine/.gitignore:1]: `.gitignore` is designed to be able to add such a file in any subdirectory within a repository, so I didn't see an issue with keeping it here. 

But I don't have a strong preference, so I'm happy to merge them into the root-level `.gitignore` file.

File: inference-engine/LICENSE
Comment (nordzilla) [reply inference-engine/LICENSE:1]: I'm not sure about this. 

@gregtatum, do you know what is standard here?

File: inference-engine/MANIFEST.in
Comment (nordzilla) [reply inference-engine/MANIFEST.in:1]: I think this is an artifact leftover from Python code that didn't get removed. 

I'll remove it.

File: inference-engine/Doxyfile.in
Comment (nordzilla) [reply inference-engine/Doxyfile.in:1]: I wasn't sure whether we would utilize it or not, so I kept it.

I guess if we need it again, we can always re-add it from upstream.

Happy to take this out for now.

File: inference-engine/BERGAMOT_VERSION
Comment (nordzilla) [reply inference-engine/BERGAMOT_VERSION:1]: This is how we currently mark the version that gets pulled into Firefox.

https://searchfox.org/mozilla-central/rev/9fa446ad77af13847a7da250135fc58b1a1bd5b9/toolkit/components/translations/bergamot-translator/moz.yaml#21-27

We likely won't need this in the future, but I don't feel comfortable removing it right now. 

I would rather remove only things that are obviously unneeded, and continue to build/remove more nuanced things as this project progresses.

File: inference-engine/README.md
Comment (nordzilla) [reply inference-engine/README.md:1]: I'll just delete the README for now, and introduce more docs additively. 

The nature of this project is a mixture between reductive and additive approaches, and I don't think we need to be splitting hairs over a vestigial README file.

File: inference-engine/examples/run-native.sh
Comment (nordzilla) [reply inference-engine/examples/run-native.sh:1]: They probably should use our models, but the purpose of this PR is to merge this repository into our repository. 

There is plenty of work still to be done. 

I would rather merge now, rather than continuing to work on a separate branch that would accumulate an even larger pull request.

**Comment by nordzilla** [comment]
Thanks for all the feedback.

I'll get this PR cleaned up further today, so that we can get this merged. 

> There's quite a lot of code. I didn't look at it much but I hope we'll be able to delete everything except the wasm bindings and related classes.

This is the ultimate goal.

The first stage is to get all of the currently used components into our repository.
The second stage is to start stripping out the HTML-related code.
The third stage is to start stripping out the segmentation-related code. 

But I need it all there to start, so that I can build up a good testing strategy, and ensure that regressions do not occur along the way. 

> There are quite a lot of small structural suggestions that should be addressed before we can merge to main. Since we decided to share documentation and lining infrastructure we should unify all this right away as it was one of the reasons why we're moving it to this repo.

I appreciate the feedback!

This is precisely why I would like to merge this into main sooner, rather than later, to ensure that we are all aligned on this kind of thing before continuing to make divergent changes on an independent branch.

File: inference-engine/LICENSE
Comment (gregtatum) [reply inference-engine/LICENSE:1]: As long as they match it should be fine to remove it.

File: inference-engine/Doxyfile.in
Comment (gregtatum) [reply inference-engine/Doxyfile.in:1]: I would prefer to follow up and keep this PR minimal.

File: inference-engine/README.md
Comment (gregtatum) [reply inference-engine/README.md:1]: I'd prefer to follow-up with clean-ups.

**Comment by nordzilla** [comment]
I made a few of the changes, resolving the relevant comments for them. 

I would prefer to follow up with the rest after this content is merged.

File: inference-engine/Doxyfile.in
Comment (eu9ene) [reply inference-engine/Doxyfile.in]: As we discussed with Erik in private I want this to be as clean as possible on the first commit for a variety of reasons. If something is not addressed here, we'll have to create an issue about this. Let me have another pass.

File: .gitmodules
Comment (eu9ene) [inline .gitmodules:38]: Let's add a TODO then

File: inference-engine/Doxyfile.in
Comment (nordzilla) [reply inference-engine/Doxyfile.in]: @eu9ene your feedback here is valuable and likely all worthwhile changes. 

But I do not understand the reservation about creating issues. Issues are cheap. They open. They get closed. They document a work item. I do not agree with the philosophy that as many things as possible should be packed into a single pull request (ostensibly for the sake of keeping the issue count low).

In fact, I would rather see more issues get created in a project than fewer issues, even if some are marked as duplicates, or closed as invalid etc. Good software engineering work is done incrementally. Issues help document that process.

**Review by eu9ene — APPROVED** [review]
It looks a lot tidier now, great job cleaning it up Erik!

I left only one comment about adding a TODO for the modules. I think we will need to unify them in consequent PRs so that the begamot module is not duplicated and we are sure we use the same version for training and inference.

**Pull request merged**
